lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1997,6 +1997,177 @@ bool swift::diagnoseArgumentLabelError(TypeChecker &TC, const Expr *expr,
+      if (auto *TE = dyn_cast<TupleExpr>(subExpr)) {
+        auto *element = TE->getElements()[TEE->getFieldNumber()];
+        return lookThroughExprsToImmediateDeallocation(element);
+      }
+      return subExpr;
+    }
+
+    if (auto *ICE = dyn_cast<ImplicitConversionExpr>(E)) {
+      E = ICE->getSubExpr();
+      continue;
+    }
+    if (auto *CE = dyn_cast<CoerceExpr>(E)) {
+      E = CE->getSubExpr();
+      continue;
+    }
+    if (auto *OEE = dyn_cast<OpenExistentialExpr>(E)) {
+      E = OEE->getSubExpr();
+      continue;
+    }
+
+    // Look through optional evaluations, we still want to diagnose on
+    // things like initializers called through optional chaining and the
+    // unwrapping of failable initializers.
+    if (auto *OEE = dyn_cast<OptionalEvaluationExpr>(E)) {
+      E = OEE->getSubExpr();
+      continue;
+    }
+    if (auto *OBE = dyn_cast<BindOptionalExpr>(E)) {
+      E = OBE->getSubExpr();
+      continue;
+    }
+    if (auto *FOE = dyn_cast<ForceValueExpr>(E)) {
+      E = FOE->getSubExpr();
+      continue;
+    }
+
+    if (auto *ATE = dyn_cast<AnyTryExpr>(E)) {
+      E = ATE->getSubExpr();
+      continue;
+    }
+    if (auto *DSBIE = dyn_cast<DotSyntaxBaseIgnoredExpr>(E)) {
+      E = DSBIE->getRHS();
+      continue;
+    }
+    return E;
+  }
+}
+
+static void diagnoseUnownedImmediateDeallocationImpl(VarDecl *varDecl,
+                                                     Expr *initExpr,
+                                                     TypeChecker &TC) {
+  auto *ownershipAttr = varDecl->getAttrs().getAttribute<OwnershipAttr>();
+  if (!ownershipAttr || ownershipAttr->isInvalid())
+    return;
+
+  // Only diagnose for non-owning ownerships such as 'weak' and 'unowned'.
+  switch (ownershipAttr->get()) {
+  case Ownership::Strong:
+    return;
+  case Ownership::Weak:
+  case Ownership::Unowned:
+  case Ownership::Unmanaged:
+    break;
+  }
+
+  // Try to find a call to a constructor.
+  initExpr = lookThroughExprsToImmediateDeallocation(initExpr);
+  auto *CE = dyn_cast<CallExpr>(initExpr);
+  if (!CE)
+    return;
+
+  auto *CRCE = dyn_cast<ConstructorRefCallExpr>(CE->getFn());
+  if (!CRCE)
+    return;
+
+  auto *DRE = dyn_cast<DeclRefExpr>(CRCE->getFn());
+  if (!DRE)
+    return;
+
+  auto *constructorDecl = dyn_cast<ConstructorDecl>(DRE->getDecl());
+  if (!constructorDecl)
+    return;
+
+  // Make sure the constructor constructs an instance that allows ownership.
+  // This is to ensure we don't diagnose on constructors such as
+  // Optional.init(nilLiteral:).
+  auto selfType = constructorDecl->getDeclContext()->getSelfTypeInContext();
+  if (!selfType->allowsOwnership())
+    return;
+
+  // This must stay in sync with
+  // diag::unowned_assignment_immediate_deallocation.
+  enum {
+    SK_Variable = 0,
+    SK_Property
+  } storageKind = SK_Variable;
+
+  if (varDecl->getDeclContext()->isTypeContext())
+    storageKind = SK_Property;
+
+  TC.diagnose(initExpr->getStartLoc(),
```

Comment:
Ideally these notes would point to the assignment, not the constructor call.

Reply:
@jrose-apple Thanks for the review!

When you say the assignment, do you mean highlighting the entire `AssignExpr`? For example:
```swift
weak var c4: C?
c4 = C()
^~~~~~~~
```

I agree that would be better, though I don't believe we can the same for arbitrary bindings, for example here are the ranges we currently highlight for a tuple pattern:

```swift
class C {}
weak var (c1, c2, c3) = (C(), C(), C())
                         ^~~  ^~~  ^~~
```
It would be ambiguous which element of the tuple we'd be referring to if we highlighted the entire thing instead.

Would it be worth special-casing a 'basic' named pattern, such that we highlight like this:

```swift
weak var c4 = C()
         ^~~~~~~~
```
instead of just on `C()`?

Reply:
No, I mean the location of the caret:

```swift
weak var c4: C?
c4 = C()
   ^ ~~~
```

That's something you can do without messing with the highlight ranges.

Reply:
@jrose-apple Ah right, I see! We can certainly do that for assignments, though for bindings it doesn't look like we keep the source loc of the equals token(s) about on the `PatternBindingDecl`. I can look into implementing that if you want, though in the interest of getting this PR moving, I think it would be best to do so in a follow up PR. Thoughts?

Reply:
Sure, that sounds fine. And yes, keeping around a SourceLoc in PatternBindingEntry for the equal sign sounds like the right thing to do.